sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -58,77 +58,126 @@ case class CreateArray(children: Seq[Expression]) extends Expression {
   }
 
   override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
-    val array = ctx.freshName("array")
-
     val et = dataType.elementType
     val evals = children.map(e => e.genCode(ctx))
     val isPrimitiveArray = ctx.isPrimitiveType(et)
-    val (preprocess, arrayData) =
-      GenArrayData.getCodeArrayData(ctx, et, children.size, isPrimitiveArray, array)
-
-    val assigns = if (isPrimitiveArray) {
-      val primitiveTypeName = ctx.primitiveTypeName(et)
-      evals.zipWithIndex.map { case (eval, i) =>
-        eval.code + s"""
-         if (${eval.isNull}) {
-           $arrayData.setNullAt($i);
-         } else {
-           $arrayData.set$primitiveTypeName($i, ${eval.value});
-         }
-       """
-      }
-    } else {
-      evals.zipWithIndex.map { case (eval, i) =>
-        eval.code + s"""
-         if (${eval.isNull}) {
-           $array[$i] = null;
-         } else {
-           $array[$i] = ${eval.value};
-         }
-       """
-      }
-    }
-    ev.copy(code =
-      preprocess +
-      ctx.splitExpressions(ctx.INPUT_ROW, assigns) +
-      s"\nfinal ArrayData ${ev.value} = $arrayData;\n",
+    val (preprocess, postprocess, arrayData, array) =
+      GenArrayData.genCodeToCreateArrayData(ctx, et, children.size, isPrimitiveArray)
+    val assigns = GenArrayData.genCodeToAssignArrayElements(
+      ctx, evals, et, isPrimitiveArray, arrayData, array, true)
+    /*
+      TODO: When we generate simpler code, we have to solve the following exception
+        https://github.com/apache/spark/pull/13909/files#r93813725
+      ev.copy(
+        code = preprocess + ctx.splitExpressions(ctx.INPUT_ROW, assigns) + postprocess
+        value = arrayData,
```

Comment:
To declare as `ArrayData` caused the same exception. Janino recognizes `ArrayData` is not assigned to `UnsafeArrayData`. Janino may keep type of each variable for simple cases.